lib/ukstore/store.c

```diff
@@ -29,3 +29,1267 @@
+
+	switch (e->type) {
+	case UK_STORE_ENTRY_TYPE(u8): {
+		__u8 val;
+
+		str = calloc(_U8_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U8_STRLEN, "%" __PRIu8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s8): {
+		__s8 val;
+
+		str = calloc(_S8_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S8_STRLEN, "%" __PRIs8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u16): {
+		__u16 val;
+
+		str = calloc(_U16_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U16_STRLEN, "%" __PRIu16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s16): {
+		__s16 val;
+
+		str = calloc(_S16_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S16_STRLEN, "%" __PRIs16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u32): {
+		__u32 val;
+
+		str = calloc(_U32_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U32_STRLEN, "%" __PRIu32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s32): {
+		__s32 val;
+
+		str = calloc(_S32_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S32_STRLEN, "%" __PRIs32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u64): {
+		__u64 val;
+
+		str = calloc(_U64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U64_STRLEN, "%" __PRIu64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s64): {
+		__s64 val;
+
+		str = calloc(_S64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S64_STRLEN, "%" __PRIs64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(uptr): {
+		__uptr val;
+
+		str = calloc(_UPTR_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _UPTR_STRLEN, "0x%" __PRIx64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(charp): {
+		char *val = NULL;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		*out  = val;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+
+int
+_uk_store_get_ncharp(const struct uk_store_entry *e, char **out, size_t maxlen)
```

Comment:
since we do not chnage the allocation of the `out` buffer, a single reference is good enough:
```c
int
uk_store_get_ncharp(const struct uk_store_entry *e, char *out, size_t maxlen)
```